zkevm-circuits/src/evm_circuit/op_execution/swap.rs

```diff
@@ -0,0 +1,482 @@
+};
+use super::{CaseAllocation, CaseConfig, OpExecutionState, OpGadget};
+use crate::util::{Expr, ToWord};
+use bus_mapping::evm::{GasCost, OpcodeId};
+use halo2::{arithmetic::FieldExt, circuit::Region, plonk::Error};
+use std::{array, convert::TryInto};
+
+#[derive(Clone, Debug)]
+struct SwapSuccessAllocation<F> {
+    case_selector: Cell<F>,
+    words: [Word<F>; 2],
+}
+
+#[derive(Clone, Debug)]
+pub struct SwapGadget<F> {
+    success: SwapSuccessAllocation<F>,
+    stack_underflow: Cell<F>, // case selector
+    out_of_gas: (
+        Cell<F>, // case selector
+        Cell<F>, // gas available
+    ),
+}
+
+impl<F: FieldExt> OpGadget<F> for SwapGadget<F> {
+    const RESPONSIBLE_OPCODES: &'static [OpcodeId] = &[
+        OpcodeId::SWAP1,
+        OpcodeId::SWAP2,
+        OpcodeId::SWAP3,
+        OpcodeId::SWAP4,
+        OpcodeId::SWAP5,
+        OpcodeId::SWAP6,
+        OpcodeId::SWAP7,
+        OpcodeId::SWAP8,
+        OpcodeId::SWAP9,
+        OpcodeId::SWAP10,
+        OpcodeId::SWAP11,
+        OpcodeId::SWAP12,
+        OpcodeId::SWAP13,
+        OpcodeId::SWAP14,
+        OpcodeId::SWAP15,
+        OpcodeId::SWAP16,
+    ];
+
+    const CASE_CONFIGS: &'static [CaseConfig] = &[
+        CaseConfig {
+            case: Case::Success,
+            num_word: 2,
+            num_cell: 0,
+            will_halt: false,
+        },
+        CaseConfig {
+            case: Case::StackUnderflow,
+            num_word: 0,
+            num_cell: 0,
+            will_halt: true,
+        },
+        CaseConfig {
+            case: Case::OutOfGas,
+            num_word: 0,
+            num_cell: 0,
+            will_halt: true,
+        },
+    ];
+
+    fn construct(case_allocations: Vec<CaseAllocation<F>>) -> Self {
+        let [success, stack_underflow, out_of_gas]: [CaseAllocation<F>; 3] =
+            case_allocations.try_into().unwrap();
+        Self {
+            success: SwapSuccessAllocation {
+                case_selector: success.selector.clone(),
+                words: success.words.try_into().unwrap(),
+            },
+            stack_underflow: stack_underflow.selector,
+            out_of_gas: (
+                out_of_gas.selector,
+                out_of_gas.resumption.unwrap().gas_available,
+            ),
+        }
+    }
+
+    fn constraints(
+        &self,
+        state_curr: &OpExecutionState<F>,
+        state_next: &OpExecutionState<F>,
+    ) -> Vec<Constraint<F>> {
+        let OpExecutionState { opcode, .. } = &state_curr;
+
+        let common_lookups = vec![Lookup::FixedLookup(
+            FixedLookup::Range16,
+            [opcode.expr() - OpcodeId::SWAP1.expr(), 0.expr(), 0.expr()],
+        )];
+
+        let num_swaped = opcode.expr() - OpcodeId::SWAP1.expr() + 1.expr();
+        let success = {
+            let SwapSuccessAllocation {
+                case_selector,
+                words,
+            } = &self.success;
+
+            let state_transition_constraints = vec![
```

Comment:
We should also constraint `stack_pointer` to be equal to next step. Otherwise the prover could choose any `stack_pointer` it wants after `SWAP*`.